fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PullUpJoinFromUnion.java

```diff
@@ -0,0 +1,361 @@
+                .withChildren(unionChild, commonChild);
+    }
+
+    private @Nullable HashMap<Plan, Set<Pair<LogicalJoin<?, ?>, Plan>>> tryToExtractCommonChild(LogicalUnion union) {
+        HashMap<Plan, Set<Pair<LogicalJoin<?, ?>, Plan>>> planCount = new HashMap<>();
+        for (Plan child : union.children()) {
+            LogicalJoin<? extends Plan, ? extends Plan> join = tryToGetJoin(child);
+            if (join == null) {
+                return null;
+            }
+            boolean added = false;
+            for (Plan plan : planCount.keySet()) {
+                LogicalPlanComparator comparator = new LogicalPlanComparator();
+                if (comparator.isLogicalEqual(join.left(), plan)) {
+                    planCount.get(plan).add(Pair.of(join, join.left()));
+                    added = true;
+                    break;
+                }
+                if (comparator.isLogicalEqual(join.right(), plan)) {
+                    planCount.get(plan).add(Pair.of(join, join.right()));
+                    added = true;
+                    break;
+                }
+            }
+
+            if (!added) {
+                planCount.put(join.left(), Sets.newHashSet(Pair.of(join, join.left())));
+                planCount.put(join.right(), Sets.newHashSet(Pair.of(join, join.right())));
+            }
+        }
+        return planCount;
+    }
+
+    private HashMap<Expression, Integer> mapSlotToIndex(LogicalUnion union) {
+        HashMap<Expression, Integer> slotToIndex = new HashMap<>();
+        for (int i = 0; i < union.children().size(); i++) {
+            Plan child = union.child(i);
+            if (union.getRegularChildOutput(i).isEmpty()) {
+                for (int slotIdx = 0; slotIdx < child.getOutput().size(); slotIdx++) {
+                    slotToIndex.put(child.getOutput().get(slotIdx), slotIdx);
+                }
+            } else {
+                for (int slotIdx = 0; slotIdx < union.getRegularChildOutput(i).size(); slotIdx++) {
+                    slotToIndex.put(union.getRegularChildOutput(i).get(slotIdx), slotIdx);
+                }
+            }
+        }
+        return slotToIndex;
+    }
+
+    // we only allow project(join) or join()
+    private @Nullable LogicalJoin<?, ?> tryToGetJoin(Plan child) {
+        if (child instanceof LogicalProject) {
+            if (!((LogicalProject<?>) child).isAllSlots()) {
+                return null;
+            }
+            child = child.child(0);
+        }
+        if (child instanceof LogicalJoin
+                && ((LogicalJoin<?, ?>) child).getJoinType().isInnerJoin()
+                && ((LogicalJoin<?, ?>) child).getOtherJoinConjuncts().isEmpty()
+                && !((LogicalJoin<?, ?>) child).isMarkJoin()) {
+            return (LogicalJoin<?, ?>) child;
+        }
+        return null;
+    }
+
+    class LogicalPlanComparator {
```

Comment:
can we extract out to a common util? Or it is just for only this rule?

Reply:
This is only a simple implementation. @seawinde will extract a more detailed implementation from mv matching